crates/soft-runtime/src/ptr.rs

```diff
@@ -0,0 +1,188 @@
+use std::{fmt::Display, marker::PhantomData, ptr::NonNull};
+
+use crate::tag::AsTagged;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum PtrTag {
+    Object = 0,
+    Number = 1,
+    Symbol = 2,
+    Pair = 3,
+}
+
+pub struct VectorNode {
+    pub size: usize,
+    pub limit: usize,
+    pub data: *mut TaggedPtr,
+}
+
+pub struct FunctionNode {
+    pub data: *mut libc::c_void,
+}
+
+pub enum ObjectNode {
+    Vector(VectorNode),
+    Function(FunctionNode),
+}
+
+pub struct SymbolNode {
+    data: &'static str,
+}
+
+pub struct PairNode {
+    head: TaggedPtr,
+    tail: TaggedPtr,
+}
+
+#[derive(Clone, Copy)]
+pub union TaggedPtr {
+    pub tag: usize,
+    pub object: NonNull<ObjectNode>,
+    pub number: isize,
```

Comment:
It would be pretty if you move `number` up to `object`, so it will align the types